File: linux/pkg/server/services/remote_registry.go
package services

import (
	"fmt"
	"sync"
)

// RemoteServer represents a registered remote server
type RemoteServer struct {
	ID         string   `json:"id"`
	Name       string   `json:"name"`
	URL        string   `json:"url"`
	Token      string   `json:"token"`
	SessionIDs []string `json:"sessionIds"`
}

// RemoteRegistry manages registered remote servers (for HQ mode)
type RemoteRegistry struct {
	mu        sync.RWMutex
	remotes   map[string]*RemoteServer
	remoteSeq int
}

// NewRemoteRegistry creates a new remote registry
func NewRemoteRegistry() *RemoteRegistry {
	return &RemoteRegistry{
		remotes: make(map[string]*RemoteServer),
	}
}

// Register adds a new remote server to the registry
func (rr *RemoteRegistry) Register(remote RemoteServer) (*RemoteServer, error) {
	rr.mu.Lock()
	defer rr.mu.Unlock()

	// Check if remote with same ID already exists
	if _, exists := rr.remotes[remote.ID]; exists {
		return nil, fmt.Errorf("remote with ID %s already registered", remote.ID)
	}

	// Check if remote with same name already exists
	for _, r := range rr.remotes {
		if r.Name == remote.Name {
			return nil, fmt.Errorf("remote with name %s already registered", remote.Name)
		}
	}

	// Initialize empty session list
	remote.SessionIDs = []string{}

	// Store the remote
	rr.remotes[remote.ID] = &remote
	rr.remoteSeq++

	return &remote, nil
}

// Unregister removes a remote server from the registry
func (rr *RemoteRegistry) Unregister(remoteID string) bool {
	rr.mu.Lock()
	defer rr.mu.Unlock()

	if _, exists := rr.remotes[remoteID]; exists {
		delete(rr.remotes, remoteID)
		return true
	}
	return false
}

// GetRemote returns a specific remote by ID
func (rr *RemoteRegistry) GetRemote(remoteID string) *RemoteServer {
	rr.mu.RLock()
	defer rr.mu.RUnlock()

	return rr.remotes[remoteID]
}

// GetRemotes returns all registered remotes
func (rr *RemoteRegistry) GetRemotes() []*RemoteServer {
	rr.mu.RLock()
	defer rr.mu.RUnlock()

	remotes := make([]*RemoteServer, 0, len(rr.remotes))
	for _, remote := range rr.remotes {
		remoteCopy := *remote
		remotes = append(remotes, &remoteCopy)
	}
	return remotes
}

// GetRemoteBySessionID finds which remote owns a specific session
func (rr *RemoteRegistry) GetRemoteBySessionID(sessionID string) *RemoteServer {
	rr.mu.RLock()
	defer rr.mu.RUnlock()

	for _, remote := range rr.remotes {
		for _, sid := range remote.SessionIDs {
			if sid == sessionID {
				return remote
			}
		}
	}
	return nil
}

// UpdateRemoteSessions updates the list of sessions for a remote
func (rr *RemoteRegistry) UpdateRemoteSessions(remoteID string, sessionIDs []string) {
	rr.mu.Lock()
	defer rr.mu.Unlock()

	if remote, exists := rr.remotes[remoteID]; exists {
		remote.SessionIDs = sessionIDs
	}
}

// AddSessionToRemote adds a session ID to a remote's session list
func (rr *RemoteRegistry) AddSessionToRemote(remoteID, sessionID string) {
	rr.mu.Lock()
	defer rr.mu.Unlock()

	if remote, exists := rr.remotes[remoteID]; exists {
		// Check if session already exists
		for _, sid := range remote.SessionIDs {
			if sid == sessionID {
				return
			}
		}
		remote.SessionIDs = append(remote.SessionIDs, sessionID)
	}
}

// RemoveSessionFromRemote removes a session ID from all remotes
func (rr *RemoteRegistry) RemoveSessionFromRemote(sessionID string) {
	rr.mu.Lock()
	defer rr.mu.Unlock()

	for _, remote := range rr.remotes {
		newSessions := []string{}
		for _, sid := range remote.SessionIDs {
			if sid != sessionID {
				newSessions = append(newSessions, sid)
			}
		}
		remote.SessionIDs = newSessions
	}
}

File: web/src/server/services/remote-registry.ts
import { isShuttingDown } from '../server.js';
import { createLogger } from '../utils/logger.js';
import chalk from 'chalk';

const logger = createLogger('remote-registry');

export interface RemoteServer {
  id: string;
  name: string;
  url: string;
  token: string;
  registeredAt: Date;
  lastHeartbeat: Date;
  sessionIds: Set<string>; // Track which sessions belong to this remote
}

export class RemoteRegistry {
  private remotes: Map<string, RemoteServer> = new Map();
  private remotesByName: Map<string, RemoteServer> = new Map();
  private sessionToRemote: Map<string, string> = new Map(); // sessionId -> remoteId
  private healthCheckInterval: NodeJS.Timeout | null = null;
  private readonly HEALTH_CHECK_INTERVAL = 15000; // Check every 15 seconds
  private readonly HEALTH_CHECK_TIMEOUT = 5000; // 5 second timeout per check

  constructor() {
    this.startHealthChecker();
    logger.debug('remote registry initialized with health check interval', {
      interval: this.HEALTH_CHECK_INTERVAL,
      timeout: this.HEALTH_CHECK_TIMEOUT,
    });
  }

  register(
    remote: Omit<RemoteServer, 'registeredAt' | 'lastHeartbeat' | 'sessionIds'>
  ): RemoteServer {
    // Check if a remote with the same name already exists
    if (this.remotesByName.has(remote.name)) {
      throw new Error(`Remote with name '${remote.name}' is already registered`);
    }

    const now = new Date();
    const registeredRemote: RemoteServer = {
      ...remote,
      registeredAt: now,
      lastHeartbeat: now,
      sessionIds: new Set<string>(),
    };

    this.remotes.set(remote.id, registeredRemote);
    this.remotesByName.set(remote.name, registeredRemote);
    logger.log(chalk.green(`remote registered: ${remote.name} (${remote.id}) from ${remote.url}`));

    // Immediately check health of new remote
    this.checkRemoteHealth(registeredRemote);

    return registeredRemote;
  }

  unregister(remoteId: string): boolean {
    const remote = this.remotes.get(remoteId);
    if (remote) {
      logger.log(chalk.yellow(`remote unregistered: ${remote.name} (${remoteId})`));

      // Clean up session mappings
      for (const sessionId of remote.sessionIds) {
        this.sessionToRemote.delete(sessionId);
      }

      this.remotesByName.delete(remote.name);
      return this.remotes.delete(remoteId);
    }
    return false;
  }

  getRemote(remoteId: string): RemoteServer | undefined {
    const remote = this.remotes.get(remoteId);
    if (!remote) {
      logger.debug(`remote not found: ${remoteId}`);
    }
    return remote;
  }

  getRemoteByUrl(url: string): RemoteServer | undefined {
    return Array.from(this.remotes.values()).find((r) => r.url === url);
  }

  getRemotes(): RemoteServer[] {
    return Array.from(this.remotes.values());
  }

  getRemoteBySessionId(sessionId: string): RemoteServer | undefined {
    const remoteId = this.sessionToRemote.get(sessionId);
    return remoteId ? this.remotes.get(remoteId) : undefined;
  }

  updateRemoteSessions(remoteId: string, sessionIds: string[]): void {
    const remote = this.remotes.get(remoteId);
    if (!remote) {
      logger.debug(`cannot update sessions: remote ${remoteId} not found`);
      return;
    }

    const oldCount = remote.sessionIds.size;

    // Remove old session mappings
    for (const oldSessionId of remote.sessionIds) {
      this.sessionToRemote.delete(oldSessionId);
    }

    // Update with new sessions
    remote.sessionIds = new Set(sessionIds);
    for (const sessionId of sessionIds) {
      this.sessionToRemote.set(sessionId, remoteId);
    }

    logger.debug(`updated sessions for remote ${remote.name}`, {
      oldCount,
      newCount: sessionIds.length,
    });
  }

  addSessionToRemote(remoteId: string, sessionId: string): void {
    const remote = this.remotes.get(remoteId);
    if (!remote) {
      logger.warn(`cannot add session ${sessionId}: remote ${remoteId} not found`);
      return;
    }

    remote.sessionIds.add(sessionId);
    this.sessionToRemote.set(sessionId, remoteId);
    logger.debug(`session ${sessionId} added to remote ${remote.name}`);
  }

  removeSessionFromRemote(sessionId: string): void {
    const remoteId = this.sessionToRemote.get(sessionId);
    if (!remoteId) {
      logger.debug(`session ${sessionId} not mapped to any remote`);
      return;
    }

    const remote = this.remotes.get(remoteId);
    if (remote) {
      remote.sessionIds.delete(sessionId);
      logger.debug(`session ${sessionId} removed from remote ${remote.name}`);
    }

    this.sessionToRemote.delete(sessionId);
  }

  private async checkRemoteHealth(remote: RemoteServer): Promise<void> {
    // Skip health checks during shutdown
    if (isShuttingDown()) {
      return;
    }

    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.HEALTH_CHECK_TIMEOUT);

      // Use the token provided by the remote for authentication
      const headers: Record<string, string> = {
        Authorization: `Bearer ${remote.token}`,
      };

      // Only check health endpoint - all remotes MUST have it
      const response = await fetch(`${remote.url}/api/health`, {
        headers,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.ok) {
        remote.lastHeartbeat = new Date();
        logger.debug(`health check passed for ${remote.name}`);
      } else {
        throw new Error(`HTTP ${response.status}`);
      }
    } catch (error) {
      // During shutdown, don't log errors or unregister remotes
      if (!isShuttingDown()) {
        logger.warn(`remote failed health check: ${remote.name} (${remote.id})`, error);
        // Remove the remote if it fails health check
        this.unregister(remote.id);
      }
    }
  }

  private startHealthChecker() {
    logger.debug('starting health checker');
    this.healthCheckInterval = setInterval(() => {
      // Skip health checks during shutdown
      if (isShuttingDown()) {
        return;
      }

      // Check all remotes in parallel
      const healthChecks = Array.from(this.remotes.values()).map((remote) =>
        this.checkRemoteHealth(remote)
      );

      Promise.all(healthChecks).catch((err) => {
        logger.error('error in health checks:', err);
      });
    }, this.HEALTH_CHECK_INTERVAL);
  }

  destroy() {
    logger.log(chalk.yellow('destroying remote registry'));
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      logger.debug('health checker stopped');
    }
  }
}
